### ISSUE TITLE
Dolby E audio track element contains multiple tracks

### ISSUE THREAD

**Issue by plondino** [MediaArea/MediaInfoLib#1846]
I have a file that has a Dolby E stream. The XML payload groups the two tracks into one track element.  I don't see an easy way to retrieve the fact that there are two tracks other than by string parsing the slash between the ID string.  It also makes the type order or stream kind position one less than the actual relative audio track number (i.e. the 16th track has StreamKindPos = 15).  

`<ID>9 / 10</ID>`

Am I missing anything that would explicitly tell me that the dolby stream takes up two tracks? It seems like I have to implicitly know that a Dolby E stream is two tracks with one channel each, and re-number accordingly, in order to enumerate the tracks correctly e.g. for a transcoder.

Your thoughts on this are appreciated.

[dolbye_mediainfo.xml.txt](https://github.com/MediaArea/MediaInfo/files/12504541/dolbye_mediainfo.xml.txt)

**Comment by JeromeMartinez** [comment]
> I don't see an easy way to retrieve the fact that there are two tracks other than by string parsing the slash between the ID string. [...] Am I missing anything that would explicitly tell me that the dolby stream takes up two tracks?

There is no explicit field for that, but we could add one for tests in MediaConch.
Note that there could be more than 2 streams for an audio content e.g. a Dolby ED2 content may be on 4 streams.

>  It also makes the type order or stream kind position one less than the actual relative audio track number (i.e. the 16th track has StreamKindPos = 15).

This becomes complicated there, we didn't define enough `StreamKindPos` for such case. Initially it is the more or less internal position in the list of audio tracks i.e. a content, which was matching the position of the audio stream i.e. the physical order in the file.
And it would be more or less random, because it depends on the count of "services" in the Dolby E content e.g. if you have 3 services in the Dolby E content, StreamKindPos of the 16th stream becomes 17.

Changing the behavior now may break some scripts, so if there is a need of "stream order" per track kind (because for a stream order per file, there is already `StreamOrder`, which is usable with FFmpeg e.g. " -map 0:17" rather than " - map 0:a:15" in the case of your file).
Do you need the stream order per track type for your transcoder?

**Comment by plondino** [comment]
Thanks for the quick response.

> Do you need the stream order per track type for your transcoder?

Yes, we basically rely on MediaInfo to tell us where the physical tracks are in the media, especially for audio.  The stream order would break if e.g. for some reason, there was a text track in between the audio tracks, correct?

**Comment by plondino** [comment]
> There is no explicit field for that, but we could add one for tests in MediaConch.
Note that there could be more than 2 streams for an audio content e.g. a Dolby ED2 content may be on 4 streams.

Yes, for now I am just hard coding. The fact that I know Dolby E takes up two channels but it would be much better for media info to report that.  But in addition to the number of channels, we really need a way to know how many tracks make up the stream e.g. for Dolby E is it one track two channels or two tracks one channel each.

**Comment by JeromeMartinez** [comment]
> The stream order would break if e.g. for some reason, there was a text track in between the audio tracks, correct?

The `StreamOrder` field is for all streams/tracks, so if there is a track not audio between the audio tracks, it will increment the field.
Your encoder has not possibility to use `StreamOrder` (` -map 0:17` with FFmpeg) and needs a type (video, audio, text...) specific  field (` - map 0:a:15`  with FFmpeg), right?

**Comment by JeromeMartinez** [comment]
> But in addition to the number of channels, we really need a way to know how many tracks make up the stream e.g. for Dolby E is it one track two channels or two tracks one channel each.

You can compute if by analyzing the ID, the ID is in the form "A / B-C", with A & B the track ID so count of stream for this track is count of " / " + 1, and C is the position of this track in the stream si count of tracks in a stream is max of C + 1.
But I guess that it is better if I show them directly in the API...

**Comment by plondino** [comment]
I am basically working on exactly the above as a workaround but yes, in a future release, it would be good to have MediaInfo enumerate all physical tracks in the file in some universal way.

**Comment by plondino** [comment]
> Your encoder has not possibility to use StreamOrder ( -map 0:17 with FFmpeg) and needs a type (video, audio, text...) specific field ( - map 0:a:15 with FFmpeg), right?

Unfortunately, we deal with several encoders both free and commercial, some of which support channel numbers, some support absolute tracks, and some only support relative track numbers.

**Comment by JeromeMartinez** [comment]
@plondino 

Currently we have that:
StreamKindID: Position of the logical track (service) of this track kind. Counting starts at 0
StreamOrder: Position of the logical track (service) whatever is the track kind. Counting starts at 0

This matches e.g FFmpeg map option, either ` -map 0:[StreamOrder]` or ` -map 0:a:[StreamKindID]` when there is only one logical track (service) per physical track (no span over several physical tracks).

Proposal of addition in order to handle the cases with several services per phsyical track and/or services spanned over several physical tracks :
PhysicalOrder: Position(s) of the physical track(s) of this track kind. Counting starts at 0
PhysicalKindOrder: Position(s) of the physical track(s) whatever is the track kind. Counting starts at 0

In practice new MediaInfo fields would be same as old ones when there is only one physical track and one service per stream.

For all MediaInfo fields (pos, order, id, ...), a generic rule (can not easily be changed now due to historic reasons and changing that now would break compatibility with lot of scripts with already implemented MediaInfo fields, and we need to think more for another form in order to be supported for every output i.e. text output, XML, JSON, ...  and having a different form depending on the field would be even more complex to explain) for complex forms:
The global value is sub values separated by "-", each sub value represent a level in the content, and is itself separated by " / " if a logical track (service) is spanned over several physical tracks.
Example: "0 / 1-2" for PhysicalOrder means that the logical track is as position 2 in the stream and that this stream is spanned over physical tracks 0 and 1; for Dolby E it would be the MediaInfo track shows metadata of the 3rd Dolby E program of a Dolby E stream spanned over physical tracks 0 and 1.
This leads for e.g. 1 video + 1 Dolby E stream having 3 programs + 1 Dolby E stream having 1 program to 5 MediaInfo tracks (1 video and 4 audio) with the following values:
StreamKindID: "0", "0", "0", "0", "1" (no Dolby E program because it was not implemented) 
StreamOrder: "0", "1-0", "1-1", "1-2", "2-0", "2-1", "2-2" (no Dolby E program because it was not implemented)
PhysicalOrder: "0", "0 / 1-0", "0 / 1-1", "0 / 1-2", "2 / 3" (no Dolby E program if only 1 program) 
PhysicalKindOrder: "0", "1 / 2-0", "1 / 2-1", "1 / 2-2", "3 / 4" (no Dolby E program if only 1 program) 
Note: I am hesitating to add the Dolby E program order, maybe not useful, or to write it even if there is only 1 Dolby E program.

If you "unpack" (erase from the dash, split values by " / ") the values this matches e.g FFmpeg map option, either ` -map 0:[PhysicalOrder first value] -map 0:[PhysicalOrder second value]` or ` -map 0:a:[PhysicalKindOrder first value] -map 0:a:[PhysicalKindOrder second value]` when there is only one logical track (service) per physical track (no span over several physical tracks).
Note that you must take care that is is not always 2 physical tracks, Dolby ED2 may span over up to 8 physical tracks.

Would it fit all your needs?

> some of which support channel numbers

What is the definition of "channel" here, especially for Dolby E? a non Dolby E aware decoder sees 2 PCM channels when a Dolby E aware decoder sees between 1 and 8 channels.
Whatever is the definition, such position can be retrieved from other MediaInfo fields.

**Comment by plondino** [comment]
I think the addition of a `PhysicalKind` and `PhysicalKindOrder` would solve my problem, if you needed to keep the StreamOrder and StreamKindID the same.   Should we call it `PhysicalKindID` to match `StreamKindID`? 

I have a few questions though about the current state - in the examples that I have, `StreamOrder`, and `StreamKindID` are integers which don't use the subkind dash notation you have above - are you planning to change the semantics of StreamOrder / StreamKindID as in your examples here? 

_StreamOrder: "0", "1-0", "1-1", "1-2", "2-0", "2-1", "2-2" (no Dolby E program because it was not implemented)_

The main issue I have is that the those fields increase monotonically so they throw off the ordering of actual tracks in the file, and I have to iterate and parse and renumber all the tracks manually using the ID field to refer to a track by its position in any system that doesn't know how to decode Dolby E - I have no access to the track position on the track itself.

Just to make sure I understand, I would expect the following with the existing usage of `StreamOrder` and `StreamKind` plus the new `PhysicalOrder` and `PhysicalKindOrder`

Track 1: Video
Track 2: PCM
Track 3: Dolby E (Track 1) Program 0
Track 4: Dolby E (Track 2) Program 0
Track 5: PCM

ID: 1, 2, 3 / 4, 5
StreamKindID: 0, 0, 1, 2
StreamOrder: 0, 1, 2, 3
PhysicalOrder: 0, 1, 2 / 3, 4
PhysicalKindOrder: 0, 0, 1 / 2, 3

If I had two Dolby E tracks with 3 programs it would be: 

Track 1: Video
Track 2: PCM
Track 3: Dolby E (Track 1) Program 0, Program 1, Program 2
Track 4: Dolby E (Track 2) Program 0, Program 1, Program 2
Track 5: PCM

ID: 1, 2, 3 / 4-1, 3 / 4-2, 5
StreamKindID: 0, 0, 1, 2, 3, 4
StreamOrder: 0, 1, 2, 3, 4, 5
PhysicalOrder: 0, 1, 2 / 3-0, 2 / 3-1, 2 / 3-2, 4
PhysicalKindOrder: 0, 0, 1 / 2-0, 1 / 2-1, 1 / 2-2, 3

**Comment by plondino** [comment]
The new fields could also be useful for embedded text tracks, where even though there are multiple streams, they all live in the same MXF track - right now this has to be parsed out of the ID field.

**Comment by JeromeMartinez** [comment]
> The main issue I have is that the those fields increase monotonically so they throw off the ordering of actual tracks in the file, and I have to iterate and parse and renumber all the tracks manually using the ID field to refer to a track by its position in any system that doesn't know how to decode Dolby E - I have no access to the track position on the track itself.

Issue is that a tool aware about Dolby E would merge the physical tracks and would show 1 logical track only for 2 physical tracks, when a tool not aware of Dolby E would show 2 logical tracks only for 2 physical tracks, so my guess is that if you don't need `StreamOrder` as it is expected, someone else could need it, or you in the future depending of the tools you try to configure. So better to have both.

Let's see, I implement this way and we see if it is good for all your use cases.

**Comment by JeromeMartinez** [comment]
> The new fields could also be useful for embedded text tracks, where even though there are multiple streams, they all live in the same physical MXF or QuickTime track - right now this has to be parsed out of the ID field.

Right, but for another issue ticket :-p.

**Comment by JeromeMartinez** [comment]
@plondino It looks I overreacted about this issue, and that `StreamOrder` is enough.
Also there is not actual need of the second part (after "-") up to now.
I cleaned up `StreamOrder` and added the 2nd stream to the field when it is Dolby E.

For the example above, it becomes:
StreamOrder: 0, 1, 2 / 3, 2 / 3, 2 / 3, 4

Let me know if you miss something after such update (except the [ANC `StreamOrder`](https://github.com/MediaArea/MediaInfo/issues/673), another topic and will be fixed too soon).
